src/DBus.vala

```diff
@@ -105,10 +130,43 @@ public class Notifications.Server : Object {
     ) throws DBusError, IOError {
         var id = (replaces_id != 0 ? replaces_id : ++id_counter);
 
+        if (replaces_id != 0) {
+            action_group.remove_actions (replaces_id);
+        }
+
         if (hints.contains (X_CANONICAL_PRIVATE_SYNCHRONOUS)) {
             send_confirmation (app_icon, hints);
         } else {
-            var notification = new Notifications.Notification (app_name, app_icon, summary, body, actions, hints);
+            var notification = new Notification (app_name, app_icon, summary, body, hints);
+            notification.buttons = new GenericArray<Notification.Button?> (actions.length / 2);
+            var actions_list = new GenericArray<string> (5);
```

Comment:
Where does the 5 come from?

Reply:
arbitrary number, it's the pre-allocated space in the array, so we don't end reallocating it after every validated action.

I belive both GNotifications and KNotifications limits the actions to 3, but there can be applications using the interface direclly adding more, i know that KDEConnect can send 5 actions with a spotify player notification. so i used it.

Reply:
ok, this got removed in the refactor. no more arbitrary number.